output/kafka/client.go

```diff
@@ -0,0 +1,436 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kafka
+
+import (
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/eapache/go-resiliency/breaker"
+
+	"github.com/elastic/beats/v7/libbeat/common/fmtstr"
+	"github.com/elastic/beats/v7/libbeat/outputs/codec"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
+)
+
+type Client struct {
+	log           *logp.Logger
+	hosts         []string
+	topic         *fmtstr.EventFormatString
+	key           *fmtstr.EventFormatString
+	index         string
+	codec         codec.Codec
+	config        sarama.Config
+	mux           sync.Mutex
+	done          chan struct{}
+	recordHeaders []sarama.RecordHeader
+
+	asyncProducer sarama.AsyncProducer
+	wg            sync.WaitGroup
+
+	//observer outputs.Observer             TODO: what to do with observers?
+	//topic    outil.Selector               TODO: Work out how to do selectors in the new world.
+
+}
+
+type MsgRef struct {
+	client *Client
+	count  int32
+	total  int
+	failed []*messages.Event // TODO: Need to know how to deal with failed events
```

Comment:
Well, my thoughts are that retry and failure behavior in the shipper is already a big standalone question that needs discussion and ideally a design document :-) Right now in Beats the retry / failure options have common elements but also vary between outputs (as one example, in Kafka the default timeout for client-to-broker communication is higher than the default timeout for broker-to-broker).

The Beats Kafka output already breaks some of our error handling abstraction because its error handling is inherently tied to our Kafka library, but the common error handling config fields are inherently tied to the Beats pipeline. We've had to do special fixes for the Kafka output because the two different error handling mechanisms don't talk to each other very well and they used to get into infinite loops because of it.

So it makes sense to have some common config baseline, but we need to pin down what the actual contract is for shipper outputs. The tentative plan as of earlier this year was that outputs would implement retries themselves but that we'd try to provide API support to make it easy (even if it's just standalone helper routines covering "the simple case"). But we need to separate the overall functional requirements (which we have at least a general idea of) from the specific API responsibilities of an output (which as far as I know we haven't started on yet).

Reply:
+1 on design doc.  I think the disk queue adds some extra edge cases.   For example, do we want to try and preserve the number of failed attempts in between shutdowns?

Reply:
As @faec mentions, there are some complications around retries in the existing kafka output - from what I can tell (despite the comment stating to the contrary), the beats kafka output [sets max retries and the back off function](https://github.com/elastic/beats/blob/main/libbeat/outputs/kafka/config.go#L261-L267) on the sarama library to have that perform retries when communicating to the server, but also calls [batch.RetryEvents](https://github.com/elastic/beats/blob/main/libbeat/outputs/kafka/client.go#L399) for events that are not determined to be unretryable (https://github.com/elastic/beats/blob/main/libbeat/outputs/kafka/client.go#L370-L375)

This PR carries over this retry mechanism, setting the number of retries and a back off function in the library config (and will return a list of "failed" events from the `publish` method), but it is unclear at the moment whether that is appropriate or sufficient. I add my +1 on a design doc here to better understand requirements and the responsibilities of the outputs.

Reply:
I'll write an issue for tracking this as well and see if we can come up with something reasonable quickly.